### PR TITLE
Add support for custom events

### PR DESCRIPTION
**Pull request by tobiaslins**
```js
import analytics from "@vercel/analytics"

analytics.track("Signup", {
  location: "footer"
})

```

### CODE REVIEW

File: packages/web/src/generic.ts [hunk 1]
```diff
@@ -34,3 +34,29 @@ export const inject = (
...
+    window.va?.('track', {
+      name: event,
+      data: props,
+    });
+  } catch (err) {
+    if (err instanceof Error && isDevelopment()) {
+      // eslint-disable-next-line no-console
+      console.error(err);
+    }
+  }
```
Comment (chriswdmr) [inline packages/web/src/generic.ts:61]: Let's use name here as it's already confusing what's an event what's the name etc.

```suggestion
  name: string,
  properties?: Record<string, AllowedPropertyValues>,
): void => {
  if (!properties) {
    window.va?.('track', { name });
    return;
  }

  try {
    const props = parseProperties(properties, {
      strip: !isDevelopment(),
    });

    window.va?.('track', {
      name,
      data: props,
    });
  } catch (err) {
    if (err instanceof Error && isDevelopment()) {
      // eslint-disable-next-line no-console
      console.error(err);
    }
  }
```

File: packages/web/src/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import { parseProperties } from './utils';
+
+describe('utils', () => {
+  describe('parse properties, strip (production)', () => {
```
Comment (chriswdmr) [inline packages/web/src/utils.test.ts:4]: Let's create a dedicated `parse properties` describe block, which contains subsequent describe blocks.

File: packages/web/src/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    it('all allowed', () => {
+      const properties = {
+        all: true,
+        should: 'pass',
+        are: true,
+        is: null,
```
Comment (chriswdmr) [inline packages/web/src/utils.test.ts:10]: Could we have a better testing object? Maybe `string: 'some-string', boolean: false`

File: packages/web/src/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      };
+
+      expect(properties).toEqual(parseProperties(properties, { strip: true }));
```
Comment (chriswdmr) [inline packages/web/src/utils.test.ts:13]: Please move that outside the expect into a dedicated line

File: packages/web/src/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+          string: 'some-string',
+          boolean: true,
+          nullable: null,
+        };
+        const parsed = parseProperties(properties, {
+          strip: false,
+        });
+
+        expect(properties).toEqual(parsed);
+      });
```
Comment (chriswdmr) [inline packages/web/src/utils.test.ts:47]: Should this be in a different `describe` as it's not throwing?

**Review by chriswdmr — COMMENTED** [review]
Nice! One thing: please add some tests for `generic.ts` and `react.tsx` so we test the actual `track` function in both cases

File: packages/web/src/utils.test.ts (hunk 5 above)
Comment (tobiaslins) [reply packages/web/src/utils.test.ts:47]: it tests the case where it would throw (and not strip) in case there is an error. but its not which is fine!

Comment (tobiaslins) [reply packages/web/src/utils.test.ts:47]: I only tested `test` once in `generic.test.ts` as the react export is the same function exported.

File: packages/web/src/generic.test.ts [hunk 6]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
...
+      if (!window.vaq) throw new Error('window.vaq is not defined');
+
+      expect(window.vaq[0]).toEqual([
+        'track',
+        {
+          name: 'my event',
+        },
+      ]);
+    });
+    it('with custom data', () => {
```
Comment (dglsparsons) [inline packages/web/src/generic.test.ts:65]: Your lack of whitespace deeply upsets me.

File: packages/web/src/generic.test.ts [hunk 7]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
...
+    it('strip nested object', () => {
+      process.env = {
+        NODE_ENV: 'production',
+      };
+      inject();
+
+      track('custom event', {
+        string: 'string',
+        number: 1,
+        // @ts-expect-error ignore for test
```
Comment (dglsparsons) [inline packages/web/src/generic.test.ts:98]: Can we cast as `any` to avoid this?

**Review by dglsparsons — APPROVED** [review]

File: packages/web/src/generic.test.ts [hunk 8]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
     });
   });
 });
+
+describe('track  custom events', () => {
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:43]:
```suggestion
describe('track custom events', () => {
```

Comment (chriswdmr) [inline packages/web/src/generic.test.ts:43]: Let's also please adds some tests for react to have it always in sync

**Review by chriswdmr — COMMENTED** [review]
Looks nice, let's add the missing tests for react and some whitespaces on the tests as Doug already pointed out

File: packages/web/src/generic.test.ts (hunk 7 above)
Comment (tobiaslins) [reply packages/web/src/generic.test.ts:98]: @dglsparsons another rule prevents us from doing this

File: packages/web/src/react.test.tsx [hunk 9]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
       expect(script).toHaveAttribute('defer');
     });
   });
+
+  describe('track  custom events', () => {
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:50]:
```suggestion
  describe('track custom events', () => {
```

File: packages/web/src/react.test.tsx [hunk 10]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
...
+        if (!window.vaq) throw new Error('window.vaq is not defined');
+
+        expect(window.vaq[0]).toEqual([
+          'track',
+          {
+            name: 'my event',
+          },
+        ]);
+      });
+      it('with custom data', () => {
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:72]:
```suggestion
      });
      
      it('with custom data', () => {
```

File: packages/web/src/react.test.tsx [hunk 11]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
       expect(script).toHaveAttribute('defer');
     });
   });
+
+  describe('track  custom events', () => {
+    beforeEach(() => {
+      // reset the internal queue before every test
+      window.vaq = [];
+    });
+    describe('queue custom events', () => {
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:55]:
```suggestion
    });
    
    describe('queue custom events', () => {
```

File: packages/web/src/react.test.tsx [hunk 12]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
...
+              number: 1,
+            },
+          },
+        ]);
+      });
+
+      it('strip nested object', () => {
+        process.env = {
+          NODE_ENV: 'production',
+        };
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:99]: That's not working as it has no effect – why do we need it?
We introduced the `mode` property for this reason.

File: packages/web/src/generic.test.ts [hunk 13]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
     });
   });
 });
+
+describe('track custom events', () => {
+  beforeEach(() => {
+    // reset the internal queue before every test
+    window.vaq = [];
+  });
+  describe('queue custom events', () => {
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:48]:
```suggestion
  });
  
  describe('queue custom events', () => {
```

File: packages/web/src/generic.test.ts [hunk 14]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
...
+            number: 1,
+          },
+        },
+      ]);
+    });
+
+    it('strip nested object', () => {
+      process.env = {
+        NODE_ENV: 'production',
+      };
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:92]: See other comment for using `process.env`

File: packages/web/src/react.test.tsx [hunk 15]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
...
+    });
+    describe('queue custom events', () => {
+      it('event name only', () => {
+        render(<Analytics mode="production" />);
+
+        track('my event');
+
+        expect(window.vaq).toBeDefined();
+
+        if (!window.vaq) throw new Error('window.vaq is not defined');
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:63]: I think this is redundant as we already have an `expect` for that on line 61

File: packages/web/src/react.test.tsx [hunk 16]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
...
+        render(<Analytics mode="production" />);
+
+        track('custom event', {
+          string: 'string',
+          number: 1,
+        });
+
+        expect(window.vaq).toBeDefined();
+
+        if (!window.vaq) throw new Error('window.vaq is not defined');
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:82]: I think this is redundant as we already have an expect for that on line 80

File: packages/web/src/react.test.tsx [hunk 17]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
...
+          number: 1,
+          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+          nested: {
+            object: '',
+          } as any,
+        });
+
+        expect(window.vaq).toBeDefined();
+
+        if (!window.vaq) throw new Error('window.vaq is not defined');
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:113]: I think this is redundant as we already have an expect for that on line 111

File: packages/web/src/utils.test.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+      it('should throw an error for arrayProp and objectProp', () => {
+        const properties = {
+          string: 'some-string',
+          arrayProp: [],
+          objectProp: {},
+        };
+
+        expect(() => {
+          parseProperties(properties, { strip: false });
```
Comment (chriswdmr) [inline packages/web/src/utils.test.ts:57]: Let's use a variable like in the other test above

File: packages/web/src/utils.test.ts [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    describe('throw error', () => {
+      it('should allow all properties', () => {
+        const properties = {
+          number: 10,
+          string: 'some-string',
+          boolean: true,
+          nullable: null,
+        };
+        const parsed = parseProperties(properties, {
```
Comment (chriswdmr) [inline packages/web/src/utils.test.ts:42]:
```suggestion
        };
        
        const parsed = parseProperties(properties, {
```

File: packages/web/src/generic.test.ts [hunk 20]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
...
+        number: 1,
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+        nested: {
+          object: '',
+        } as any,
+      });
+
+      expect(window.vaq).toBeDefined();
+
+      if (!window.vaq) throw new Error('window.vaq is not defined');
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:106]: I think this is redundant as we already have an expect for that on line 104

File: packages/web/src/generic.test.ts [hunk 21]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
...
+      inject();
+
+      track('custom event', {
+        string: 'string',
+        number: 1,
+      });
+
+      expect(window.vaq).toBeDefined();
+
+      if (!window.vaq) throw new Error('window.vaq is not defined');
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:75]: I think this is redundant as we already have an expect for that on line 73

File: packages/web/src/generic.test.ts [hunk 22]
```diff
@@ -39,3 +39,82 @@ describe('inject', () => {
...
+  });
+  describe('queue custom events', () => {
+    it('event name only', () => {
+      inject();
+
+      track('my event');
+
+      expect(window.vaq).toBeDefined();
+
+      if (!window.vaq) throw new Error('window.vaq is not defined');
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:56]: I think this is redundant as we already have an expect for that on line 54

Comment (dglsparsons) [reply packages/web/src/generic.test.ts:56]: It probably satisfies typescript though.

File: packages/web/src/react.test.tsx (hunk 10 above)
Comment (dglsparsons) [reply packages/web/src/react.test.tsx:72]: `it` should define the behaviour too, `it('should allow custom data to be tracked', () => {` maybe?

File: packages/web/src/react.test.tsx [hunk 23]
```diff
@@ -46,4 +46,83 @@ describe('<Analytics />', () => {
...
+            name: 'custom event',
+            data: {
+              string: 'string',
+              number: 1,
+            },
+          },
+        ]);
+      });
+
+      it('strip nested object', () => {
```
Comment (dglsparsons) [inline packages/web/src/react.test.tsx:96]:
```suggestion
      it('should strip data for nested objects', () => {
```

File: packages/web/src/generic.test.ts (hunk 22 above)
Comment (dglsparsons) [reply packages/web/src/generic.test.ts:56]: I think so. I'd argue the `toBeDefined` may be redundant, but at least we'll get a nice test failure that way.

File: packages/web/src/generic.test.ts (hunk 20 above)
Comment (tobiaslins) [reply packages/web/src/generic.test.ts:106]: @chriswdmr it satisfies typescript

File: packages/web/src/utils.test.ts (hunk 19 above)
Comment (tobiaslins) [reply packages/web/src/utils.test.ts:42]: if we bug about those new lines, we should just add eslint/prettier rules for that.

File: packages/web/src/utils.test.ts (hunk 18 above)
Comment (tobiaslins) [reply packages/web/src/utils.test.ts:57]: its a closure thats required for catching errors. we could move it out but I don't see a reason to think about about a name + we're not reusing it.

File: packages/web/src/react.test.tsx (hunk 12 above)
Comment (tobiaslins) [reply packages/web/src/react.test.tsx:99]: I just saw that I used `isDevelopment` in the `track` function, thats why.
We need to think about a solution on how I can access `mode` without manually passing it to `track`.

File: packages/web/src/utils.test.ts (hunk 18 above)
Comment (chriswdmr) [reply packages/web/src/utils.test.ts:57]: Oh, true! Sorry, makes sense!

File: packages/web/src/react.test.tsx (hunk 12 above)
Comment (chriswdmr) [reply packages/web/src/react.test.tsx]: For that we have `getMode`. It's also using `isDevelopment` under the hood

Comment (chriswdmr) [reply packages/web/src/react.test.tsx]: I believe we have to store the `mode` on a global level, so subsequent calls such as `track` can read that value that was provided on `inject`. That should allow us to have `track` and other functions without the additional effort of providing a `mode` property individually.

1. `inject` is called with `mode`
2. we store the `mode` on `va` globally
3. `track` reads mode from global object as it was already set by `inject`

Comment (tobiaslins) [reply packages/web/src/react.test.tsx]: @chriswdmr `inject` now sets `window.vam` to the mode that is set. Track later reads this to know if its prod or not!

File: packages/web/src/generic.ts [hunk 24]
```diff
@@ -11,6 +11,8 @@ export const inject = (
 
   const mode = getMode(props.mode);
 
+  window.vam = mode;
```
Comment (chriswdmr) [inline packages/web/src/generic.ts:14]: Let's create `setMode` which does set `window.vam`

I would suggest that we create `isProduction` and `isDevelopment` – both are reading from `window.vam` so we're consistent every time we try to identify the correct environment.

This should also eliminate the need of manually reading `window.vam` on `track`

Comment (tobiaslins) [reply packages/web/src/generic.ts]: I renamed `isDevelopment` to `detectDevelopment` and `getMode` to `detectMode` 
`setMode` and `getMode` are now just wrapping `window.vam` and `isProduction` and `isDevelopment` is reading from `getMode`

File: packages/web/src/generic.ts [hunk 25]
```diff
@@ -9,28 +16,26 @@ export const inject = (
 ): void => {
   if (!isBrowser()) return;
 
-  const mode = getMode(props.mode);
-
-  window.vam = mode;
+  const mode = detectMode(props.mode);
+  setMode(mode);
```
Comment (chriswdmr) [inline packages/web/src/generic.ts:20]: Could we please combine those things? What about `setMode(props.mode)` only as we use `isDevelopment` and `isProduction`

File: packages/web/src/utils.ts [hunk 26]
```diff
@@ -1,20 +1,67 @@
...
   return typeof window !== 'undefined';
 }
 
-export function isDevelopment(): boolean {
+export function detectDevelopment(): boolean {
   if (typeof process === 'undefined') return false;
   return (
     process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test'
   );
 }
```
Comment (chriswdmr) [inline packages/web/src/utils.ts:12]: This logic can be moved to `isDevelopment`

File: packages/web/src/react.test.tsx [hunk 27]
```diff
@@ -46,4 +46,82 @@ describe('<Analytics />', () => {
       expect(script).toHaveAttribute('defer');
     });
   });
+
+  describe('track custom events', () => {
```
Comment (chriswdmr) [inline packages/web/src/react.test.tsx:50]: Just noticed that we're missing two describe blocks for `'in development mode'` and `'in production mode'` to test the separate modes

File: packages/web/src/generic.test.ts [hunk 28]
```diff
@@ -39,3 +39,83 @@ describe('inject', () => {
     });
   });
 });
+
+describe('track custom events', () => {
```
Comment (chriswdmr) [inline packages/web/src/generic.test.ts:43]: Just noticed that we're missing two describe blocks for `'in development mode'` and `'in production mode'` to test the separate modes

File: packages/web/src/react.test.tsx (hunk 27 above)
Comment (tobiaslins) [reply packages/web/src/react.test.tsx:50]: did it once in generic. not needed to just copy it

**Review by chriswdmr — APPROVED** [review]

File: apps/nextjs/package.json [hunk 29]
```diff
@@ -9,7 +9,7 @@
     "test:e2e:production": "NEXT_PUBLIC_ANALYTICS_MODE=production pnpm --filter nextjs... build && playwright test production"
   },
   "dependencies": {
-    "@vercel/analytics": "workspace:^0.1.5",
+    "@vercel/analytics": "workspace:../../packages/web",
```
Comment (dglsparsons) [inline apps/nextjs/package.json:12]: ? Why's this changed?

Comment (tobiaslins) [reply apps/nextjs/package.json:12]: we should always test against latest package? tests where failing and after I changed that it worked again

Comment (dglsparsons) [reply apps/nextjs/package.json:12]:
```suggestion
    "@vercel/analytics": "workspace:*",
```

I think i see the issue - using `^0.1.5` means it'll only use the local version while semver matches `^0.1.5`. Bumping the package could break that. 

Instead, we want this test app to _always_ use the workspace package. 
So we should have `workspace:*`.

**Review by dglsparsons — APPROVED** [review]

**Review by dglsparsons — APPROVED** [review]

**Comment by fabianuribe** [comment]
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

**Pull request merged**